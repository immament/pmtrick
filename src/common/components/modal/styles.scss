.pmt-modal.modal {
    display: block !important;

    text-shadow: none;
    color: #333333;
}
