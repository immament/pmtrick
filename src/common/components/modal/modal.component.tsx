import './styles.scss';

import React, { ReactNode } from 'react';
import ReactDOM from 'react-dom';

const modalRoot: HTMLElement = document.body;

export type ModalProps = {
    onClose?: (e: React.MouseEvent<HTMLButtonElement>) => void;
    show: boolean;
    title: string;
    size?: 'lg' | 'sm';
    className?: string;
};
export class Modal extends React.Component<ModalProps> {
    private element: Element;

    constructor(props: ModalProps) {
        super(props);
        this.element = document.createElement('div');
    }

    componentDidMount(): void {
        modalRoot.appendChild(this.element);
    }

    componentWillUnmount(): void {
        modalRoot.removeChild(this.element);
    }

    private onClose = (e: React.MouseEvent<HTMLButtonElement>): void => {
        this.props.onClose && this.props.onClose(e);
    };

    private modalClassName(): string {
        return `modal-dialog ${this.props.size ? 'modal-' + this.props.size : ''} ${this.props.className || ''}`;
    }

    render(): ReactNode {
        if (!this.props.show) {
            return null;
        }
        return ReactDOM.createPortal(
            <div className="pmt-modal modal" id="modal">
                <div className={this.modalClassName()} role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <button
                                type="button"
                                className="close"
                                data-dismiss="modal"
                                aria-label="Close"
                                onClick={this.onClose}
                            >
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 className="modal-title">{this.props.title}</h4>
                        </div>
                        <div className="modal-body">{this.props.children}</div>
                        <div className="modal-footer">
                            <button
                                type="button"
                                className="btn btn-default"
                                data-dismiss="modal"
                                onClick={this.onClose}
                            >
                                Close
                            </button>
                        </div>
                    </div>
                </div>
            </div>,
            this.element,
        );
    }
}
