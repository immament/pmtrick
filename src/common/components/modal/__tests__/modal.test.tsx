import React, { ReactNode, ReactPortal } from 'react';
import ReactDOM from 'react-dom';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

import { Modal } from '../modal.component';

describe('Modal', () => {
    const oldCreatePortal = ReactDOM.createPortal;
    beforeAll(() => {
        (ReactDOM.createPortal as unknown) = jest.fn((element, _node) => {
            return element;
        });
    });

    afterAll(() => {
        // (ReactDOM.createPortal as jest.Mock).mockClear();
        ReactDOM.createPortal = oldCreatePortal;
    });
    it('component renders', () => {
        ReactDOM.createPortal = (children: ReactNode): ReactPortal => children as ReactPortal; // for components that use Portal

        const handleClose = (_e: React.MouseEvent<Element>): void => {
            // TODO
        };

        const tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = renderer
            .create(<Modal show={true} title="Sample title" onClose={handleClose} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
