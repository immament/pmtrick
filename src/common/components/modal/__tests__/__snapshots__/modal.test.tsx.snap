// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders 1`] = `
<div
  className="pmt-modal modal"
  id="modal"
>
  <div
    className="modal-dialog "
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            Ã—
          </span>
        </button>
        <h4
          className="modal-title"
        >
          Sample title
        </h4>
      </div>
      <div
        className="modal-body"
      />
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-default"
          data-dismiss="modal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
